### ISSUE TITLE
Graph Neural Networks

### ISSUE THREAD

**Issue by stergiosba** [patrick-kidger/equinox#702]
Greetings Patrick,

I was wondering, what are your thoughts when it comes to Graph Neural Networks? Would a GNNlayer be implemented as a module for Equinox or an advanced example for the documentation?

 Sorry if this has been brought up before, I couldn't find a similar issue/question.

**Comment by patrick-kidger** [comment]
I probably wouldn't add it to the core library, but I'd be happy to have that as one of our examples!

**Comment by mvcisback** [comment]
@patrick-kidger @stergiosba 

I'm trying my hand at writing a GATv2 (https://arxiv.org/abs/2105.14491) implementation loosely based on https://nn.labml.ai/graphs/gatv2/index.html

*note*: I don't have multiheaded implemented, optional weight sharing, or dropout. Although I think multiheaded and droppout are better done outside the layer via `vmap` anyway...

I plan to test it works using the dictionary lookup problem:
https://github.com/mvcisback/dict_lookup_mpnn_problem

Any feedback would be appreciated and I'd be happy to provide it as an example if I can get it up to snuff.

**Comment by patrick-kidger** [comment]
> Any feedback would be appreciated and I'd be happy to provide it as an example if I can get it up to snuff

That would be awesome! In particular I'm always happy to have more examples.

**Comment by mvcisback** [comment]
@patrick-kidger 

I have the start of the endeavor here. Hope to update with some training code soon.

https://github.com/mvcisback/gatv2_eqx

**Comment by stergiosba** [comment]
@mvcisback 
I have added an issue in your project.

**Comment by mvcisback** [comment]
> Oh awesome! I'm enjoying tracking the progress of this :)
> 
> As for performance -- I'm guessing you mean that things are recompiling as the number of nodes changes? Yup, padding is the standard solution for this. (And thankfully very easy to do for GNNs, it's basically just adding a few totally disconnected nodes.)

Yeah. That's what I ended up doing in the example problem.

https://github.com/mvcisback/gatv2_eqx/blob/main/examples/train_dict_lookup.py

It ended up looking a bit uglier, but now as far as I can tell it works!

I'm gonna fiddle around a bit to better understand what the jit system will let me get away with.
